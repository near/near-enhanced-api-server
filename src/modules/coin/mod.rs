use actix_web::{HttpResponse, error};
use actix_web_validator::{Error, PathConfig};
use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

#[derive(serde::Serialize)]
pub struct ValidationErrorJsonPayload {
    pub message: String,
    pub fields: Vec<String>,
}

/// Custom error handler
impl From<&validator::ValidationErrors> for ValidationErrorJsonPayload {
    fn from(error: &validator::ValidationErrors) -> Self {
        ValidationErrorJsonPayload {
            message: "Validation error".to_owned(),
            fields: error.field_errors().iter().map(|(field, _)| field.to_string()).collect(),
        }
    }
}

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app
        .app_data(PathConfig::default().error_handler(|err, _| {
            let json_error = match &err {
                Error::Validate(error) => ValidationErrorJsonPayload::from(error),
                _ => ValidationErrorJsonPayload { message: err.to_string(), fields: Vec::new() },
            };
            error::InternalError::from_response(err, HttpResponse::Conflict().json(json_error)).into()
        }))
        .service(
        web::resource("/accounts/{account_id}/coins/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins")
            .route(web::get().to(resources::get_coin_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}")
            .route(web::get().to(resources::get_coin_balances_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}/history")
            .route(web::get().to(resources::get_coin_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_contract_metadata)),
    );
}
