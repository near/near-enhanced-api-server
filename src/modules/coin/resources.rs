use paperclip::actix::{
    api_v2_operation,
    web::{self, Json},
};

use super::{data_provider, schemas};
use crate::{db_helpers, errors, modules, types};

#[api_v2_operation(tags(Coins))]
/// Get user's NEAR balance
///
/// This endpoint returns the NEAR balance of the given `account_id`
/// at the given `block_timestamp_nanos`/`block_height`.
pub async fn get_near_balance(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::NearBalanceResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    Ok(Json(
        data_provider::get_near_balance(&pool, &block, &request.account_id.0).await?,
    ))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin balances
///
/// This endpoint returns all the countable coin balances (including NEAR, FTs)
/// of the given `account_id`, at the given `block_timestamp_nanos`/`block_height`.
/// todo it's not great, we can't keep sorting + pagination when we add MT here
/// NEAR balance always goes first, then the balances are sorted alphabetically by `contract_account_id`.
///
/// You can also provide `limit` (max 100) and `after_contract_account_id`.
/// `after_contract_account_id` is the way to paginate through the values.
///
/// Some pagination notes: leave `after_contract_account_id` empty at your first request;
/// then, take `contract_account_id` of the last provided item in your previous response.
///
/// **Limitations**
/// * We are in the process of supporting Multi Token balances.
pub async fn get_coin_balances(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::CoinBalancesResponse>> {
    let mut limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let mut balances: Vec<schemas::Coin> = vec![];
    balances.push(
        data_provider::get_near_balance(&pool, &block, &request.account_id.0)
            .await?
            .into(),
    );
    limit -= 1;

    if limit > 0 {
        let ft_balances = &mut data_provider::get_coin_balances(
            &pool,
            &rpc_client,
            &block,
            &request.account_id.0,
            limit,
        )
        .await?;
        balances.append(ft_balances);
    }

    Ok(Json(schemas::CoinBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin balances by contract
///
/// This endpoint returns all the countable coin balances of the given `account_id`,
/// for the given `contract_account_id` and `block_timestamp_nanos`/`block_height`.
/// For FT contracts, the response has only 1 item in the list.
/// For MT contracts, there could be several balances (MT support is still under development).
/// todo change interface for ft/mt
///
/// **Limitations**
/// * We are in the process of supporting Multi Token balances.
pub async fn get_coin_balances_by_contract(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceByContractRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::CoinBalancesResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native balance, please use NEAR (uppercase)".to_string(),
        )
        .into());
    }
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let balances = data_provider::get_coin_balances_by_contract(
        &rpc_client,
        &block,
        &request.contract_account_id.0,
        &request.account_id.0,
    )
    .await?;

    Ok(Json(schemas::CoinBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's NEAR history
///
/// This endpoint returns the history of native NEAR operations,
/// for the given `account_id`, going from the latest to the earliest events.
/// For the next page, use `after_event_index` and pass there the index from the last event from the previous page.
///
/// **Limitations**
/// We are in the process of collecting the data. We support only last 3 months history for now.
pub async fn get_near_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_near_history(
            &pool_balances.pool,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin history by contract
///
/// This endpoint returns the history of coin operations (FT, other standards)
/// for the given `account_id`, `contract_id`, going from the latest to the earliest events.
/// For the next page, use `after_event_index` and pass there the index from the last event from the previous page.
///
/// **Limitations**
/// * If the history does not match with the balances from RPC, we do not support the history (see Trusted/Inconsistent Contracts section below).
/// * We are in the process of supporting Multi Token history.
///
/// **Trusted/Inconsistent Contracts**
/// The contracts produce [the balance changing events](https://nomicon.io/Standards/Tokens/FungibleToken/Event).
/// The contract is marked as inconsistent if its events do not match with the balance taken from RPC.
/// The contract is marked as trusted otherwise.
/// More info could be found [here](https://github.com/near/near-indexer-events#my-contract-produces-eventstheres-a-custom-legacy-logic-for-my-contract-but-the-enhanced-api-still-ignores-me-why).
pub async fn get_coin_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::HistoryRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native history, please use NEAR (uppercase)".to_string(),
        )
        .into());
    }

    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_coin_history(
            &pool,
            &pool_balances.pool,
            &rpc_client,
            &request.contract_account_id.0,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get FT contract metadata
///
/// This endpoint returns the metadata for a given FT contract and `timestamp`/`block_height`.
///
/// **Limitations**
/// * For now, we support only FT contracts that implement the Events NEP standard.
///   We are working on a solution to support other FT contracts, including `wrap.near` and bridged tokens.
pub async fn get_ft_contract_metadata(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::ContractMetadataRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::FtContractMetadataResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;

    Ok(Json(schemas::FtContractMetadataResponse {
        metadata: data_provider::get_ft_contract_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}
