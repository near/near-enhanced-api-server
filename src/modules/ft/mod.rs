use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/accounts/{account_id}/balances/FT")
            .route(web::get().to(resources::get_ft_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/FT/{contract_account_id}")
            .route(web::get().to(resources::get_ft_balance_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/FT/{contract_account_id}/history")
            .route(web::get().to(resources::get_ft_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_metadata)),
    );
}
