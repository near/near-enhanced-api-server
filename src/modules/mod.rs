use crate::{db_helpers, errors, types};

pub(crate) mod coin;
pub(crate) mod nft;

pub(crate) async fn check_account_exists(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    account_id: &near_primitives::types::AccountId,
    block_height: u64,
) -> crate::Result<()> {
    let request = near_jsonrpc_client::methods::query::RpcQueryRequest {
        block_reference: near_primitives::types::BlockId::Height(block_height).into(),
        request: near_primitives::views::QueryRequest::ViewAccount {
            account_id: account_id.clone(),
        },
    };
    for _ in 0..5 {
        match rpc_client.call(&request).await {
            Err(near_jsonrpc_client::errors::JsonRpcError::ServerError(
                near_jsonrpc_client::errors::JsonRpcServerError::HandlerError(
                    near_jsonrpc_client::methods::query::RpcQueryError::UnknownAccount { .. },
                ),
            )) => {
                return Err(errors::ErrorKind::InvalidInput(format!(
                    "account_id {} does not exist at block_height {}",
                    account_id, block_height
                ))
                .into())
            }
            Err(err) => {
                tracing::warn!(target: crate::LOGGER_MSG, "Checking account existence via JSON RPC failed with: {:?}. Re-trying immediatelly", err);
                continue;
            }
            Ok(response) => {
                if let near_jsonrpc_primitives::types::query::QueryResponseKind::ViewAccount(_) =
                    response.kind
                {
                    return Ok(());
                } else {
                    tracing::warn!(target: crate::LOGGER_MSG, "Checking account existence returned invalid response: {:?}. Re-trying immediatelly", response);
                    continue;
                }
            }
        }
    }
    Err(errors::ErrorKind::InternalError(format!(
        "could not check if account_id {} exists after several attemps",
        account_id
    ))
    .into())
}

pub(crate) async fn check_and_get_history_pagination_params(
    pool: &sqlx::Pool<sqlx::Postgres>,
    pagination_params: types::query_params::HistoryPaginationParams,
) -> crate::Result<types::query_params::HistoryPagination> {
    types::query_params::check_limit(pagination_params.limit)?;
    let pagination = types::query_params::Pagination::from(pagination_params);
    // if pagination_params.after_block_height.is_some() && pagination_params.after_timestamp_nanos.is_some() {
    //     return Err(errors::ErrorKind::InvalidInput(
    //         "Both block_height and block_timestamp_nanos found. Please provide only one of values"
    //             .to_string(),
    //     )
    //         .into());
    // }
    // TODO PHASE 2 take the block from pagination_params
    let block = db_helpers::get_last_block(pool).await?;
    Ok(types::query_params::HistoryPagination {
        block_height: block.height,
        block_timestamp: block.timestamp,
        limit: pagination.limit,
    })
}

#[cfg(test)]
mod tests {
    use crate::db_helpers;

    pub(crate) async fn init_db() -> sqlx::Pool<sqlx::Postgres> {
        dotenv::dotenv().ok();
        let db_url = &std::env::var("DATABASE_URL").expect("failed to get database url");

        sqlx::PgPool::connect(db_url)
            .await
            .expect("failed to connect to the database")
    }

    pub(crate) fn init_rpc() -> near_jsonrpc_client::JsonRpcClient {
        dotenv::dotenv().ok();
        let rpc_url = &std::env::var("RPC_URL").expect("failed to get RPC url");
        let connector = near_jsonrpc_client::JsonRpcClient::new_client();
        connector.connect(rpc_url)
    }

    pub(crate) fn get_block() -> db_helpers::Block {
        db_helpers::Block {
            timestamp: 1655571176644255779,
            height: 68000000,
        }
    }
}
