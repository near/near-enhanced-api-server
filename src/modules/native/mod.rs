use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/accounts/{account_id}/balances/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    );
}
