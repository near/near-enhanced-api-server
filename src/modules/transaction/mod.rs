use paperclip::actix::web;

// mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/transaction/{transaction_hash}")
            .route(web::get().to(resources::get_transaction_by_tx_hash)),
    )
    .service(
        web::resource("/transaction/receipt/{receipt_id}")
            .route(web::get().to(resources::get_transaction_by_receipt)),
    )
    .service(
        web::resource("/transactions/{account_id}")
            .route(web::get().to(resources::get_transactions_by_account)),
    )
    .service(
        web::resource("/transactions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_transactions_by_account_on_contract)),
    )
    .service(
        web::resource("/receipts/{transaction_hash}")
            .route(web::get().to(resources::get_receipts_by_tx_hash)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}")
            .route(web::get().to(resources::get_action_receipts_by_account)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_action_receipts_by_account_on_contract)),
    );
}
