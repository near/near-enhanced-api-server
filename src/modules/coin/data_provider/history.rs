use std::str::FromStr;

use crate::modules::coin;
use crate::{db_helpers, errors, types};

// TODO PHASE 2 pagination by artificial index added to balance_changes
pub(crate) async fn get_near_history(
    balances_pool: &sqlx::Pool<sqlx::Postgres>,
    account_id: &near_primitives::types::AccountId,
    pagination: &types::query_params::HistoryPagination,
) -> crate::Result<Vec<coin::schemas::HistoryItem>> {
    let query = r"
        SELECT
            involved_account_id,
            delta_nonstaked_amount + delta_staked_amount delta_balance,
            absolute_nonstaked_amount + absolute_staked_amount balance,
            cause,
            status,
            block_timestamp block_timestamp_nanos
        FROM balance_changes
        WHERE affected_account_id = $1 AND block_timestamp < $2::numeric(20, 0)
        ORDER BY block_timestamp DESC
        LIMIT $3::numeric(20, 0)
    ";

    let history_info = db_helpers::select_retry_or_panic::<super::models::NearHistoryInfo>(
        balances_pool,
        query,
        &[
            account_id.to_string(),
            pagination.block_timestamp.to_string(),
            pagination.limit.to_string(),
        ],
    )
    .await?;

    let mut result: Vec<coin::schemas::HistoryItem> = vec![];
    for history in history_info {
        result.push(history.try_into()?);
    }
    Ok(result)
}

// TODO PHASE 2 pagination by artificial index added to assets__fungible_token_events
// TODO PHASE 2 change RPC call to DB call by adding absolute amount values to assets__fungible_token_events
// TODO PHASE 2 make the decision about separate FT/MT tables or one table. Pagination implementation depends on this
pub(crate) async fn get_coin_history(
    pool: &sqlx::Pool<sqlx::Postgres>,
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: &near_primitives::types::AccountId,
    account_id: &near_primitives::types::AccountId,
    pagination: &types::query_params::HistoryPagination,
) -> crate::Result<Vec<coin::schemas::HistoryItem>> {
    // this is temp solution before we make changes to the DB
    let mut last_balance = super::balance::get_ft_balance_by_contract(
        rpc_client,
        contract_id.clone(),
        account_id.clone(),
        pagination.block_height,
    )
    .await?;
    let metadata = coin::schemas::CoinMetadata::from(
        super::metadata::get_ft_contract_metadata(
            rpc_client,
            contract_id.clone(),
            pagination.block_height,
        )
        .await?,
    );

    let account_id = account_id.to_string();
    let query = r"
        SELECT
            -- blocks.block_height,
            blocks.block_timestamp,
            assets__fungible_token_events.amount::numeric(45, 0),
            assets__fungible_token_events.event_kind::text cause,
            CASE WHEN execution_outcomes.status IN ('SUCCESS_VALUE', 'SUCCESS_RECEIPT_ID') THEN 'SUCCESS'
                ELSE 'FAILURE'
            END status,
            assets__fungible_token_events.token_old_owner_account_id old_owner_id,
            assets__fungible_token_events.token_new_owner_account_id new_owner_id
        FROM assets__fungible_token_events
            JOIN blocks ON assets__fungible_token_events.emitted_at_block_timestamp = blocks.block_timestamp
            JOIN execution_outcomes ON assets__fungible_token_events.emitted_for_receipt_id = execution_outcomes.receipt_id
        WHERE emitted_by_contract_account_id = $1
            AND (token_old_owner_account_id = $2 OR token_new_owner_account_id = $2)
            AND emitted_at_block_timestamp <= $3::numeric(20, 0)
        ORDER BY emitted_at_block_timestamp desc
        LIMIT $4::numeric(20, 0)
    ";
    let history_info = db_helpers::select_retry_or_panic::<super::models::CoinHistoryInfo>(
        pool,
        query,
        &[
            contract_id.to_string(),
            account_id.clone(),
            pagination.block_timestamp.to_string(),
            pagination.limit.to_string(),
        ],
    )
    .await?;

    let mut result: Vec<coin::schemas::HistoryItem> = vec![];
    for db_info in history_info {
        let mut delta: i128 = types::numeric::to_i128(&db_info.amount)?;
        let balance = last_balance;
        // TODO PHASE 2 maybe we want to change assets__fungible_token_events also to affected/involved?
        let involved_account_id = if account_id == db_info.old_owner_id {
            delta = -delta;
            types::account_id::extract_account_id(&db_info.new_owner_id)?
        } else if account_id == db_info.new_owner_id {
            types::account_id::extract_account_id(&db_info.old_owner_id)?
        } else {
            return Err(
                errors::ErrorKind::InternalError(
                    format!("The account {} should be sender or receiver ({}, {}). If you see this, please create the issue",
                            account_id, db_info.old_owner_id, db_info.new_owner_id)).into(),
            );
        };

        if db_info.status == "SUCCESS" {
            // TODO PHASE 2 this strange error will go away after we add absolute amounts to the DB
            if (last_balance as i128) - delta < 0 {
                return Err(errors::ErrorKind::InternalError(format!(
                    "Balance could not be negative: account {}, contract {}",
                    account_id, contract_id
                ))
                .into());
            }
            last_balance = ((last_balance as i128) - delta) as u128;
        }

        result.push(coin::schemas::HistoryItem {
            cause: db_info.cause.clone(),
            involved_account_id: involved_account_id.map(|id| id.into()),
            delta_balance: delta.into(),
            balance: balance.into(),
            coin_metadata: metadata.clone(),
            block_timestamp_nanos: types::numeric::to_u64(&db_info.block_timestamp)?.into(),
            // block_height: types::numeric::to_u64(&db_info.block_height)?.into(),
            status: db_info.status,
        });
    }
    Ok(result)
}

impl TryFrom<super::models::NearHistoryInfo> for coin::schemas::HistoryItem {
    type Error = errors::Error;

    fn try_from(info: super::models::NearHistoryInfo) -> crate::Result<Self> {
        let involved_account_id: Option<types::AccountId> =
            if let Some(account_id) = info.involved_account_id {
                Some(near_primitives::types::AccountId::from_str(&account_id)?.into())
            } else {
                None
            };
        Ok(Self {
            involved_account_id,
            delta_balance: types::numeric::to_i128(&info.delta_balance)?.into(),
            balance: types::numeric::to_u128(&info.balance)?.into(),
            cause: info.cause,
            status: info.status,
            coin_metadata: super::get_near_metadata(),
            block_timestamp_nanos: types::numeric::to_u64(&info.block_timestamp_nanos)?.into(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::tests::*;

    #[tokio::test]
    async fn test_near_history() {
        let block = get_block();
        // Using the other pool because we have this table at the other DB
        dotenv::dotenv().ok();
        let url_balances =
            &std::env::var("DATABASE_URL_BALANCES").expect("failed to get database url");
        let pool = sqlx::PgPool::connect(url_balances)
            .await
            .expect("failed to connect to the balances database");
        let account = near_primitives::types::AccountId::from_str("vasya.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_near_history(&pool, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_near_history_with_failed_receipts() {
        let block = db_helpers::Block {
            timestamp: 1618591017607373869,
            height: 34943083
        };
        // Using the other pool because we have this table at the other DB
        dotenv::dotenv().ok();
        let url_balances =
            &std::env::var("DATABASE_URL_BALANCES").expect("failed to get database url");
        let pool = sqlx::PgPool::connect(url_balances)
            .await
            .expect("failed to connect to the balances database");
        let account = near_primitives::types::AccountId::from_str("zubkowi.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_near_history(&pool, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_coin_history() {
        let pool = init_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("usn").unwrap();
        let account = near_primitives::types::AccountId::from_str("pushxo.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_coin_history(&pool, &rpc_client, &contract, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_coin_history_with_failed_receipts() {
        let pool = init_db().await;
        let rpc_client = init_rpc();
        let block = db_helpers::Block {
            timestamp: 1651062637353692535,
            height: 64408633
        };
        let contract = near_primitives::types::AccountId::from_str("sweat_token_testing.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("intmainreturn0.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_coin_history(&pool, &rpc_client, &contract, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }
}
