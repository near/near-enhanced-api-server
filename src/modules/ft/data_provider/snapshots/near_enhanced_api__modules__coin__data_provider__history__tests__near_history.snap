---
source: src/modules/coin/data_provider/history.rs
expression: balance
---
Ok(
    [
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                101623970103823687500,
            ),
            balance: U128(
                1720232195281062100155460,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238577144947703,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                21805236558567385300904,
            ),
            balance: U128(
                1720130571310958276467960,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238576023448127,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                132895885775025917792770,
            ),
            balance: U128(
                1698325334752390891167056,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238574957099510,
            ),
        },
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zomland.near",
                ),
            ),
            delta_balance: I128(
                0,
            ),
            balance: U128(
                1565429448977364973374286,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238573809891921,
            ),
        },
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zomland.near",
                ),
            ),
            delta_balance: I128(
                -167086398855095326781174,
            ),
            balance: U128(
                1565429448977364973374286,
            ),
            cause: "TRANSACTION",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238572443234931,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                101623970103823687500,
            ),
            balance: U128(
                1732515847832460300155460,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238555768202334,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                9411594535554956060890,
            ),
            balance: U128(
                1732414223862356476467960,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238554622402591,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                57286270361355272541374,
            ),
            balance: U128(
                1723002629326801520407070,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238553414981139,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                17557003937520464358912,
            ),
            balance: U128(
                1665716358965446247865696,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238552045091840,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                68473053959772436884896,
            ),
            balance: U128(
                1648159355027925783506784,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1655238550653634874,
            ),
        },
    ],
)
